### PR TITLE
Use fp32 in optimizer step

### PR DESCRIPTION
**Pull request by zyksir**
## Motivation

In Mixed Precision Training, we should use fp32 to do `optimizer.step`; both Megatron and DeepSpeed are using this way. 

## Modifications

**please DONOT  Merge this PR for now, I haven't fully tested this commit yet.**

Add `BF16Optimizer`, make a copy of fp32 params, every time when we want to do optimizer.step:
1. We copy the bf16 param grad to fp32 param grad
2. We do optimizer.step on fp32 param
3. We copy fp32 param weight to bf16 param weight

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by zhyncs** [comment]
@zyksir please fix conflicts

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[submodule "third_party/sglang"]
+	path = third_party/sglang
+	url = https://github.com/sgl-project/sglang.git
+	branch = v0.4.10.post2
```
Comment (zhyncs) [inline .gitmodules]: can we use v0.5.1?

Comment (zhyncs) [reply .gitmodules]: and why do we need to install from source, can we use pypi version?

**Review by zhyncs — APPROVED** [review]
other parts LGTM

File: .gitmodules (hunk 1 above)
Comment (zyksir) [reply .gitmodules]: fixed and now using v0.5.1

File: requirements.txt [hunk 2]
```diff
@@ -11,3 +12,4 @@ psutil
 numpy
 accelerate
 pydantic
+sglang==v0.5.1
```
Comment (zhyncs) [inline requirements.txt]: sglang[all]==0.5.1 ?

Comment (zyksir) [reply requirements.txt]: done
